File: web_version/pmugg_files/network/primalVolume.js
// Boundary is a face. Interior should be null.
ms.primalVolume = function(type) {
	this.boundary = [];
	this.interior = [];
	this.type = type;

	this.id = ms.counter.add('primalVolume');
};

ms.primalVolume.prototype.getType     = function(){	return this.type; };
ms.primalVolume.prototype.getBoundary = function(){	return this.boundary[0]; };
ms.primalVolume.prototype.getInterior = function(){	return this.interior[0]; };

// Turns don't matter for volumes.
ms.primalVolume.prototype.getTurns = function(){	return 0; };
ms.primalVolume.prototype.setTurns = function(turns){ };

ms.primalVolume.prototype.copy = function(){
	return new ms.primalVolume(this.type);
};

ms.primalVolume.prototype.export = function(){
	var i = this.getInterior();
	var b = this.getBoundary();
	return {
		boundary: b ? b.getNetwork().faceIndex(b) : -1,
		type: this.type,
	};
};

ms.primalVolume.prototype.import = function(boundNet, types, json){
	var boundaryB = json.boundary >= 0 ? boundNet.boundary.getFaces()[json.boundary] : null;
	ms.boundNet.connectBoundary(this, boundaryB);
	// var interiorB = json.interior >= 0 ? boundNet.interior.getFaces()[json.interior] : null;
	// ms.boundNet.connectInterior(this, interiorB);
	this.type = json.type;
};

ms.primalVolume.prototype.boundaryString = function() {
	var outerHalfs = this.getBoundary().getOuterHalfEdges();
	var result = '';
	outerHalfs.forEach((half) => {
		result += half.getEdge().getPrimal().boundaryString();
	});
	return result;
};

ms.primalVolume.prototype.print = function() {
	console.log(this.boundaryString());
	this.boundary.print();
};

ms.counter.register('primalVolume');

File: cpp_version/pmugg/network/primal_volume.cpp
#include "primal_volume.h"
#include "dual_object.h"
#include "bound_net.h"
#include "types.h"
#include "half_edge.h"
#include "face.h"
#include "util.h"

namespace ms {

int PrimalVolume::nextId = 0;

PrimalVolume::PrimalVolume(int volumeType)
    : type(volumeType)
    , id(nextId++) {}

std::string PrimalVolume::boundaryString() const {
    if (!getBoundary()) return "";

    std::string result;
    auto* face = getBoundary();
    auto outerHalfs = face->getOuterHalfEdges();

    for (auto* half : outerHalfs) {
        if (auto* edge = half->getEdge()) {
            if (auto* primal = edge->getPrimal()) {
                result += primal->boundaryString();
            }
        }
    }

    return result;
}

void PrimalVolume::print() const {
    std::cout << boundaryString() << std::endl;
    if (getBoundary()) {
        getBoundary()->print();
    }
}

PrimalVolume* PrimalVolume::copy() const {
    return new PrimalVolume(type);
}

void PrimalVolume::import(BoundNet* boundNet, Types* types, const Json& json) {
    // Get boundary connection
    int boundaryIndex = json["boundary"].get<int>();
    if (boundaryIndex >= 0) {
        auto* boundaryB = boundNet->getBoundary()->getFaces()[boundaryIndex];
        BoundNet::connectBoundary(this, boundaryB);
    }

    // Set type
    type = json["type"].get<int>();
}

Json PrimalVolume::export() const {
    Json json;
    
    // Export boundary connection
    auto* b = getBoundary();
    json["boundary"] = b ? b->getNetwork()->faceIndex(b) : -1;
    
    // Export type
    json["type"] = type;
    
    return json;
}

} // namespace ms 